backends/cuda/ceed-cuda-basis.c

```diff
@@ -0,0 +1,278 @@
+// Copyright (c) 2017-2018, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory. LLNL-CODE-734707.
+// All Rights reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+#include <ceed-impl.h>
+#include "../include/ceed.h"
+#include "ceed-cuda.h"
+
+static const char *basiskernels = QUOTE(
+extern "C" __global__ void interp(const CeedInt nelem, const int transpose, const CeedScalar * __restrict__ interp1d, const CeedScalar * __restrict__ u, CeedScalar *__restrict__ v) {
+  const CeedInt i = threadIdx.x;
+
+  __shared__ CeedScalar s_mem[BASIS_Q1D * BASIS_P1D + 2 * BASIS_BUF_LEN];
+  CeedScalar *s_interp1d = s_mem;
+  CeedScalar *s_buf1 = s_mem + BASIS_Q1D * BASIS_P1D;
+  CeedScalar *s_buf2 = s_buf1 + BASIS_BUF_LEN;
+  for (CeedInt k = i; k < BASIS_Q1D * BASIS_P1D; k += blockDim.x) {
+    s_interp1d[k] = interp1d[k];
+  }
+
+  const CeedInt P = transpose ? BASIS_Q1D : BASIS_P1D;
```

Comment:
Will this cover non-tensor bases as well or will that be a future modification?

Reply:
I haven't touched at all the GPU implementations except what wasn't written and the `weight` kernel. I'm working on a separate branch on a similar branch using my GPU implementations instead of @Steven-Roberts 's.
I haven't spent the time to understand the logic of this implementation, and look if it could be extended to non-tensor basis. I imagine that it could be hacked using `BASIS_DIM=1` and `BASIS_Q1D=Q`, since applying the contraction we use is a basically a matrix vector product.

Reply:
Fair. I think you are right about setting the backend BASIS_DIM=1 and BASIS_Q1D=Q. We do something similar to enable non-tensor bases for Ref and Blocked.